diffrax/_progress_bar.py

```diff
@@ -0,0 +1,174 @@
+import abc
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+from jax.experimental import io_callback
+from jaxtyping import Array, PyTree
+
+from ._custom_types import FloatScalarLike, IntScalarLike
+
+
+class AbstractProgressBar(eqx.Module):
+    @abc.abstractmethod
+    def init(self) -> PyTree[Array]:
+        """Creates a new progress bar and returns its unique identifier"""
+
+    @abc.abstractmethod
+    def step(self, state: PyTree[Array], progress: FloatScalarLike) -> PyTree[Array]:
+        """Updates the progress bar and returns the last printed progress value"""
+
+    @abc.abstractmethod
+    def close(self, state: PyTree[Array]):
+        """Closes the progress bar"""
+
+
+class NoProgressBar(AbstractProgressBar):
+    def init(self) -> None:
+        return None
+
+    def step(self, state, progress: FloatScalarLike) -> None:
+        return state
+
+    def close(self, state):
+        pass
+
+
+class _TextProgressBarState(eqx.Module):
+    progress: FloatScalarLike
+
+    def __init__(self, progress):
+        self.progress = progress
+
+
+class TextProgressBar(AbstractProgressBar):
+    minimum_increase: float
+
+    def __init__(self, minimum_increase: float = 2.0):
+        """
+        **Arguments:**
+
+        - `minimum_increase`: minimum increase of progress bar in percent
+        """
+        self.minimum_increase = minimum_increase
+
+    def init(self) -> _TextProgressBarState:
+        return _TextProgressBarState(progress=0.0)
+
+    def step(
+        self, state: _TextProgressBarState, progress: FloatScalarLike
+    ) -> _TextProgressBarState:
+        def update_aux(progress):
+            # currently
+            # jax.debug.print("{progress:.2f}%", progress=progress)
+            # does not work, so we have to round the progress ourselves
+            jax.debug.print("{}%", jnp.round(progress, 1), ordered=True)
+            return progress
+
+        # we only print if the progress has increased by at least
+        # `minimum_increase` to avoid flooding the user
+        # with too many updates
+        next_progress = jax.lax.cond(
+            progress - state.progress > self.minimum_increase,
+            update_aux,
+            lambda p: state.progress,
+            progress,
+        )
+
+        return _TextProgressBarState(progress=next_progress)
+
+    def close(self, state: _TextProgressBarState):
+        jax.debug.print("100%", ordered=True)
+
+
+class _TqdmProgressBarState(eqx.Module):
+    progress_bar_id: IntScalarLike
+    step: IntScalarLike
+
+    def __init__(self, progress_bar_id, step):
+        self.progress_bar_id = progress_bar_id
+        self.step = step
+
+
+class TqdmProgressBar(AbstractProgressBar):
+    refresh_rate: int
+
+    def __init__(self, refresh_rate: int = 20):
+        """
+        **Arguments:**
+
+        - `refresh_rate`: number of steps between refreshes
+        """
+        self.refresh_rate = refresh_rate
```

Comment:
Oh actually, and here! (For the docstring I tend to just do `Foo.__init__.__doc__ = "..."` after the class definition, see e.g. https://github.com/patrick-kidger/diffrax/blob/8aafefcd081d51240ac865662a1767ca999d2455/diffrax/_step_size_controller/adaptive.py#L711 .)